proposals/infra/INF-0005-register-types-and-diagnostics.md

```diff
@@ -130,21 +130,275 @@ is no corresponding member, a warning will be emitted that is treated as an erro
 because this behavior was permissible in legacy versions of the compiler. These warnings
 are also part of the same warning group, `LegacyConstantRegisterBinding`.
 
-`DiagnoseHLSLRegisterAttribute` will also be responsible for emitting a diagnostic
-if any other invalid register type is detected. If `DiagnoseHLSLRegisterAttribute`
-finds any critical errors, the attribute, `HLSLResourceBindingAttr`, won't be added
+`DiagnoseHLSLRegisterAttribute` will also be responsible for emitting a diagnostic 
+if any other invalid register type is detected. If `DiagnoseHLSLRegisterAttribute` 
+finds any critical errors, the attribute, `HLSLResourceBindingAttr`, won't be added 
 to the Decl, and compilation will fail. However, `DiagnoseHLSLRegisterAttribute` may
-emit some warnings and allow the attribute to be attached. In summary,
-`DiagnoseHLSLRegisterAttribute` will be responsible for analyzing the context of the
+emit some warnings and allow the attribute to be attached. In summary, 
+`DiagnoseHLSLRegisterAttribute` will be responsible for analyzing the context of the 
 decl to which the register annotation is being applied, and using the data in the
-annotation to determine what diagnostics, if any, to emit.
-`DiagnoseHLSLRegisterAttribute` will be fully responsible for halting compilation
+annotation to determine what diagnostics, if any, to emit. 
+`DiagnoseHLSLRegisterAttribute` will be fully responsible for halting compilation 
 if there is any semantic fault in the application of the register annotation.
 
-
 ## Detailed design
+Firstly, below is some sample code that would cause the proposed diagnostics to be emitted:
+```
+// UDTs with register bindings for resources that don't exist:
+//
+// - warn_hlsl_user_defined_type_missing_member
+// warning: binding type '%select{t|u|b|s|c}0' only applies to types containing '%select{srv resources|uav resources|constant buffer resources|sampler state|constant buffer data}0'
+struct Foo { float f; };
+Foo x : register(t0);
+// warning: binding type 't' only applies to types containing 'srv' resources.
+//
+
+// Mismatched register bindings:
+//
+// - err_hlsl_binding_type_mismatch
+// error: binding type '%select{t|u|b|s|c}0' only applies to '%select{srv resources|uav resources|constant buffer resources|sampler state|constant buffer data}0'
+float f : register(t0);
+// error: binding type 't' only applies to 'srv' resources.
+RWBuffer<float> f : register(c3);
+// error: binding type 'c' only applies to constant buffer data.

```

Comment:
I'm wondering if "constant buffer data" is the correct term to use here? It only applies to offsets in $Globals, doesn't it?

If you use it inside a `cbuffer` then you get a warning to use `packoffset`.  I think earlier in the document it talks about "offset into $Globals", I wonder if some of that text could make it here?

Reply:
I think the `offset` part is an implementation detail that doesn't need to be mentioned in the diagnostic, but I could totally see changing the message to:
`...only applies to $Globals constant buffer data`
Thoughts?

Reply:
I'm not sure I agree that the offset part is an implementation detail - the developer absolutely needs to know what that does!

However, for a diagnostic I think it's ok to lean on just trying to describe what it has to apply to.  

Did we find a snappy way to classify things that go into $Globals?  Is it "global data" or "numeric globals" perhaps?

Reply:
Any thoughts on resolution to this one?

Reply:
How about `default cbuffer globals`?

Reply:
TBH, I think any mention of cbuffer and $Globals is probably confusing here.  How about: 

`error: binding type 'c' only applies to variables in the global scope`